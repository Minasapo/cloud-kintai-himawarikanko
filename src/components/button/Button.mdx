import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Source,
  Story,
} from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

# Button

アプリケーション内で使用するボタンのコンポーネントです。\
具体的なユースケースが決まっているものは、ストーリーに記載されています。

<Primary />

# パラメーター

コンポーネント内で使用できるパラメーターです。\
パラメーターを変更することで上にあるコンポーネントが変化します。

<Controls />

# ストーリー

コンポーネントのユースケースを記載しています。\
既存のコンポーネントを使用する際に参考にしてください。

<Stories />
