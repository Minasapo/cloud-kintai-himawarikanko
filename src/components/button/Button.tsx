// cspell:words testid
import { Button as MuiButton } from "@mui/material";

import { Color, Variant } from "../../lib/theme";

type Size = "small" | "medium" | "large";

export interface ButtonProps {
  color?: Color;
  variant?: Variant;
  disabled?: boolean;
  size?: Size;
  label?: string;
  width?: string;
  height?: string;
  borderRadius?: string;
  onClick?: () => void;
}

const Button = ({
  color = "primary",
  variant = "contained",
  disabled = false,
  size = "medium",
  label = "Button",
  onClick = () => {},
  ...props
}: ButtonProps) => (
  <MuiButton
    color={color}
    variant={variant}
    disabled={disabled}
    size={size}
    onClick={onClick}
    style={{ ...props }}
  >
    {label}
  </MuiButton>
);

export default Button;
