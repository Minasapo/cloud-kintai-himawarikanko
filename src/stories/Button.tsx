import { Button as MuiButton, createTheme, ThemeProvider } from "@mui/material";
import "./button.css";

type Color =
  | "inherit"
  | "primary"
  | "secondary"
  | "success"
  | "error"
  | "info"
  | "warning"
  | "cancel"
  | "rest"
  | "clock_in"
  | "clock_out";
type Variant = "text" | "outlined" | "contained";
type Size = "small" | "medium" | "large";

interface ButtonProps {
  color?: Color;
  variant?: Variant;
  disabled?: boolean;
  size?: Size;
  label?: string;
  width?: string;
  height?: string;
  borderRadius?: string;
  onClick?: () => void;
}

declare module "@mui/material/styles" {
  interface Palette {
    cancel: Palette["error"];
    rest: Palette["primary"];
  }

  interface PaletteOptions {
    cancel?: PaletteOptions["error"];
    rest?: PaletteOptions["primary"];
    clock_in?: PaletteOptions["primary"];
    clock_out?: PaletteOptions["error"];
  }
}

declare module "@mui/material/Button" {
  interface ButtonPropsColorOverrides {
    cancel: true;
    rest: true;
    clock_in: true;
    clock_out: true;
  }

  interface ButtonPropsVariantOverrides {
    cancel: true;
    rest: true;
    clock_in: true;
    clock_out: true;
  }
}

const palette = {
  primary: { main: "#0FA85E", contrastText: "#fff" },
  secondary: { main: "#B33D47", contrastText: "#fff" },
  success: { main: "#0FA85E", contrastText: "#fff" },
  error: { main: "#B33D47", contrastText: "#fff" },
  warning: { main: "#B33D47", contrastText: "#fff" },
  cancel: { main: "#B33D47", contrastText: "#fff" },
  rest: { main: "#2ACEDB", contrastText: "#fff" },
  clock_in: { main: "#0FA85E", contrastText: "#fff" },
  clock_out: { main: "#B33D47", contrastText: "#fff" },
};

const theme = createTheme({
  palette,
  typography: {
    fontFamily: ["Noto Sans JP"].join(","),
  },
  components: {
    MuiCssBaseline: {
      styleOverrides: `
        @font-face {
          font-family: 'Noto Sans JP';
          font-style: bold;
          font-weight: 700;
        }`,
    },
    MuiButton: {
      variants: [
        {
          props: { variant: "contained", color: "primary" },
          style: {
            backgroundColor: palette.primary.main,
            color: palette.primary.contrastText,
            boxShadow: "none",
            border: `3px solid ${palette.primary.main}`,
            "&:hover": {
              backgroundColor: palette.primary.contrastText,
              color: palette.primary.main,
              border: `3px solid ${palette.primary.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
        {
          props: { variant: "contained", color: "cancel" },
          style: {
            backgroundColor: palette.cancel.main,
            color: palette.cancel.contrastText,
            boxShadow: "none",
            border: `3px solid ${palette.cancel.main}`,
            "&:hover": {
              backgroundColor: palette.cancel.contrastText,
              color: palette.cancel.main,
              border: "3px solid",
              boxShadow: "none",
            },
          },
        },
        {
          props: { variant: "contained", color: "rest" },
          style: {
            backgroundColor: palette.rest.main,
            color: palette.rest.contrastText,
            boxShadow: "none",
            border: `3px solid ${palette.rest.main}`,
            "&:hover": {
              backgroundColor: palette.rest.contrastText,
              color: palette.rest.main,
              border: `3px solid ${palette.rest.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
        {
          props: { variant: "outlined", color: "clock_in", size: "large" },
          style: {
            backgroundColor: palette.clock_in.contrastText,
            color: palette.clock_in.main,
            boxShadow: "none",
            border: `3px solid ${palette.clock_in.main}`,
            borderRadius: "50%",
            width: "120px",
            height: "120px",
            "&:hover": {
              backgroundColor: palette.clock_in.main,
              color: palette.clock_in.contrastText,
              border: `3px solid ${palette.clock_in.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
        {
          props: { variant: "contained", color: "clock_in", size: "large" },
          style: {
            backgroundColor: palette.clock_in.main,
            color: palette.clock_in.contrastText,
            boxShadow: "none",
            border: `3px solid ${palette.clock_in.main}`,
            borderRadius: "50%",
            width: "120px",
            height: "120px",
            "&:hover": {
              backgroundColor: palette.clock_in.contrastText,
              color: palette.clock_in.main,
              border: `3px solid ${palette.clock_in.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
        {
          props: { variant: "outlined", color: "clock_out", size: "large" },
          style: {
            backgroundColor: palette.clock_out.contrastText,
            color: palette.clock_out.main,
            boxShadow: "none",
            border: `3px solid ${palette.clock_out.main}`,
            borderRadius: "50%",
            width: "120px",
            height: "120px",
            "&:hover": {
              backgroundColor: palette.clock_out.main,
              color: palette.clock_out.contrastText,
              border: `3px solid ${palette.clock_out.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
        {
          props: { variant: "contained", color: "clock_out", size: "large" },
          style: {
            backgroundColor: palette.clock_out.main,
            color: palette.clock_out.contrastText,
            boxShadow: "none",
            border: `3px solid ${palette.clock_out.main}`,
            borderRadius: "50%",
            width: "120px",
            height: "120px",
            "&:hover": {
              backgroundColor: palette.clock_out.contrastText,
              color: palette.clock_out.main,
              border: `3px solid ${palette.clock_out.main}`,
              boxShadow: "none",
            },
            "&:disabled": {
              border: "3px solid",
            },
          },
        },
      ],
    },
  },
});

const Button = ({
  color = "primary",
  variant = "contained",
  disabled = false,
  size = "medium",
  label = "Button",

  ...props
}: ButtonProps) => (
  <ThemeProvider theme={theme}>
    <MuiButton
      color={color}
      variant={variant}
      disabled={disabled}
      size={size}
      style={{ ...props }}
    >
      {label}
    </MuiButton>
  </ThemeProvider>
);

export default Button;
