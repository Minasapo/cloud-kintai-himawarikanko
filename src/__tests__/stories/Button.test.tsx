import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";
import Button from "../../stories/Button";

describe("Button", () => {
  test("onClick", () => {
    const onClickMock = jest.fn();
    render(<Button label="label" onClick={onClickMock} />);
    userEvent.click(screen.getByRole("button"));

    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test("Disabled", () => {
    render(<Button label="label" disabled />);
    expect(screen.getByRole("button")).toBeDisabled();
  });

  test("Snapshot", () => {
    const { container } = render(<Button label="label" />);
    expect(container).toMatchSnapshot();
  });
});
