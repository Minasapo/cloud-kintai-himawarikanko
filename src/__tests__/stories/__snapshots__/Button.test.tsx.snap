// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-htqi8w-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    label
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
